src/TypeScript/TsConfig/TsConfigsLocator.cs

```diff
@@ -21,47 +21,109 @@
 using System;
 using System.Collections.Generic;
 using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Abstractions;
 using System.Linq;
 using Microsoft.VisualStudio;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
+using SonarLint.VisualStudio.Infrastructure.VS;
+using SonarLint.VisualStudio.Integration;
 
 namespace SonarLint.VisualStudio.TypeScript.TsConfig
 {
     internal interface ITsConfigsLocator
     {
         /// <summary>
-        /// Returns all the tsconfig files in the current solution.
+        /// Returns all the tsconfig files in the project directory of the given <see cref="sourceFilePath"/>.
         /// </summary>
-        IReadOnlyList<string> Locate();
+        IReadOnlyList<string> Locate(string sourceFilePath);
     }
 
     [Export(typeof(ITsConfigsLocator))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     internal class TsConfigsLocator : ITsConfigsLocator
     {
-        internal const int MaxNumberOfFiles = 1000;
+        /// <summary>
+        /// See https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.__vspropid7?view=visualstudiosdk-2019
+        /// The enum is not available in VS 2015 API.
+        /// </summary>
+        internal const int VSPROPID_IsInOpenFolderMode = -8044;
 
-        private readonly IVsUIShellOpenDocument vsUiShellOpenDocument;
+        private readonly IVsHierarchyLocator vsHierarchyLocator;
+        private readonly IVsSolution vsSolution;
+        private readonly IFileSystem fileSystem;
+        private readonly ILogger logger;
 
         [ImportingConstructor]
-        public TsConfigsLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        public TsConfigsLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsHierarchyLocator vsHierarchyLocator,
+            ILogger logger)
+            : this(serviceProvider, vsHierarchyLocator, new FileSystem(), logger)
+        {
+        }
+
+        internal TsConfigsLocator(IServiceProvider serviceProvider,
+            IVsHierarchyLocator vsHierarchyLocator,
+            IFileSystem fileSystem, 
+            ILogger logger)
         {
-            vsUiShellOpenDocument = serviceProvider.GetService(typeof(SVsUIShellOpenDocument)) as IVsUIShellOpenDocument;
+            vsSolution = serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            this.vsHierarchyLocator = vsHierarchyLocator;
+            this.fileSystem = fileSystem;
+            this.logger = logger;
+        }
+
+        public IReadOnlyList<string> Locate(string sourceFilePath)
+        {
+            var projectDirectory = GetProjectDirectory(sourceFilePath);
+
+            if (projectDirectory == null)
+            {
+                return Array.Empty<string>();
+            }
+
+            var foundTsConfigs = fileSystem
```

Comment:
Worth a comment that this behaviour matches the observed VS behaviour i.e. only the file location matters, not whether they are referenced in the project.